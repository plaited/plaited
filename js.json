{
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "plaited",
    "paths": {
      "plaited/jsx-runtime": [
        "./src/jsx/runtime.ts"
      ],
      "plaited/jsx-dev-runtime": [
        "./src/jsx/dev-runtime.ts"
      ]
    }
  }
}