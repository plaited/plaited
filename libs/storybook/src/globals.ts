import { global } from '@storybook/global';

const { window: globalWindow } = global;


declare global {
  interface Window {
    STORYBOOK_ENV: 'plaited';
  }
}

if (globalWindow) {
  globalWindow.STORYBOOK_ENV = 'plaited';
}
