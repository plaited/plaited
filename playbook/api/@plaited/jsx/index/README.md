**@plaited/jsx** ( [Readme](../README.md) \| API )

***

[Plaited Typedocs](../../../modules.md) / [@plaited/jsx](../modules.md) / index

# index

## Contents

- [Interfaces](README.md#interfaces)
- [Type Aliases](README.md#type-aliases)
- [Variables](README.md#variables)
- [Functions](README.md#functions)
- [References](README.md#references)
  - [h](README.md#h)

## Interfaces

- [AdditionalAttrs](interfaces/AdditionalAttrs.md)
- [CreateTemplate](interfaces/CreateTemplate.md)

## Type Aliases

- [Attrs](type-aliases/Attrs.md)
- [BaseAttrs](type-aliases/BaseAttrs.md)
- [Child](type-aliases/Child.md)
- [Children](type-aliases/Children.md)
- [FT](type-aliases/FT.md)
- [FunctionTemplate](type-aliases/FunctionTemplate.md)
- [Primitive](type-aliases/Primitive.md)
- [Template](type-aliases/Template.md)

## Variables

- [booleanAttrs](variables/booleanAttrs.md)
- [dataAddress](variables/dataAddress.md)
- [dataTarget](variables/dataTarget.md)
- [dataTrigger](variables/dataTrigger.md)
- [primitives](variables/primitives.md)
- [validPrimitiveChildren](variables/validPrimitiveChildren.md)
- [voidTags](variables/voidTags.md)

## Functions

- [Fragment](functions/Fragment.md)
- [classNames](functions/classNames.md)
- [createTemplate](functions/createTemplate.md)
- [css](functions/css.md)
- [memo](functions/memo.md)
- [stylesheets](functions/stylesheets.md)

## References

### h

Renames and re-exports [createTemplate](functions/createTemplate.md)

***

Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
