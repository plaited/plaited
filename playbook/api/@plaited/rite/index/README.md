**@plaited/rite** ( [Readme](../README.md) \| API )

***

[Plaited Typedocs](../../../modules.md) / [@plaited/rite](../modules.md) / index

# index

## Index

### Namespaces

- [assert](namespaces/assert/README.md)
- [test](namespaces/test/README.md)

### Classes

- [AssertionError](classes/AssertionError.md)

### Interfaces

- [Assertion](interfaces/Assertion.md)

### Functions

- [assert](functions/assert.md)
- [t](functions/t.md)
- [test](functions/test.md)

***

Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
