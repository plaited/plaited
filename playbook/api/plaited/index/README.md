**plaited** ( [Readme](../README.md) \| API )

***

[Plaited Typedocs](../../modules.md) / [plaited](../modules.md) / index

# index

## Contents

- [Interfaces](README.md#interfaces)
- [Type Aliases](README.md#type-aliases)
- [Variables](README.md#variables)
- [Functions](README.md#functions)
- [References](README.md#references)
  - [h](README.md#h)

## Interfaces

- [$](interfaces/$.md)
- [AdditionalAttrs](interfaces/AdditionalAttrs.md)
- [Connect](interfaces/Connect.md)
- [CreateTemplate](interfaces/CreateTemplate.md)
- [DevCallback](interfaces/DevCallback.md)
- [PlaitedComponentConstructor](interfaces/PlaitedComponentConstructor.md)
- [PlaitedElement](interfaces/PlaitedElement.md)
- [StateSnapshot](interfaces/StateSnapshot.md)

## Type Aliases

- [Attrs](type-aliases/Attrs.md)
- [BaseAttrs](type-aliases/BaseAttrs.md)
- [CandidateBid](type-aliases/CandidateBid.md)
- [Child](type-aliases/Child.md)
- [Children](type-aliases/Children.md)
- [ComponentFunction](type-aliases/ComponentFunction.md)
- [Detail](type-aliases/Detail.md)
- [FT](type-aliases/FT.md)
- [Feedback](type-aliases/Feedback.md)
- [FunctionTemplate](type-aliases/FunctionTemplate.md)
- [Log](type-aliases/Log.md)
- [Message](type-aliases/Message.md)
- [ParameterIdiom](type-aliases/ParameterIdiom.md)
- [PendingBid](type-aliases/PendingBid.md)
- [PlaitProps](type-aliases/PlaitProps.md)
- [Primitive](type-aliases/Primitive.md)
- [RequestIdiom](type-aliases/RequestIdiom.md)
- [RuleSet](type-aliases/RuleSet.md)
- [RulesFunc](type-aliases/RulesFunc.md)
- [RunningBid](type-aliases/RunningBid.md)
- [SelectedMessage](type-aliases/SelectedMessage.md)
- [SelectorMod](type-aliases/SelectorMod.md)
- [Send](type-aliases/Send.md)
- [SnapshotMessage](type-aliases/SnapshotMessage.md)
- [Strategy](type-aliases/Strategy.md)
- [Template](type-aliases/Template.md)
- [Trigger](type-aliases/Trigger.md)
- [TriggerArgs](type-aliases/TriggerArgs.md)

## Variables

- [booleanAttrs](variables/booleanAttrs.md)
- [dataAddress](variables/dataAddress.md)
- [dataTarget](variables/dataTarget.md)
- [dataTrigger](variables/dataTrigger.md)
- [primitives](variables/primitives.md)
- [validPrimitiveChildren](variables/validPrimitiveChildren.md)
- [voidTags](variables/voidTags.md)

## Functions

- [Component](functions/Component.md)
- [Fragment](functions/Fragment.md)
- [bProgram](functions/bProgram.md)
- [classNames](functions/classNames.md)
- [createTemplate](functions/createTemplate.md)
- [createTemplateElement](functions/createTemplateElement.md)
- [css](functions/css.md)
- [loop](functions/loop.md)
- [memo](functions/memo.md)
- [stylesheets](functions/stylesheets.md)
- [sync](functions/sync.md)
- [thread](functions/thread.md)
- [useIndexedDB](functions/useIndexedDB.md)
- [useMain](functions/useMain.md)
- [useMessenger](functions/useMessenger.md)
- [useStore](functions/useStore.md)

## References

### h

Renames and re-exports [createTemplate](functions/createTemplate.md)

***

Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
