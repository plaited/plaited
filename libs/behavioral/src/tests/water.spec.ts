import { test, expect } from'@jest/globals'
import { bProgram, DevCallback } from '../index.js'

test('Add hot water 3 times', () => {
  const actual: string[] = []
  const { addThreads, thread, sync, trigger, feedback } = bProgram()
  addThreads({
    addHot: thread(
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } })
    ),
  })
  feedback({
    hot() {
      actual.push('hot')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual([ 'hot', 'hot', 'hot' ])
})

test('Add hot/cold water 3 times', () => {
  const actual: string[] = []
  const { addThreads, thread, sync, trigger, feedback } = bProgram()
  addThreads({
    addHot: thread(
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } })
    ),
    addCold: thread(
      sync({ request: { type: 'cold' } }),
      sync({ request: { type: 'cold' } }),
      sync({ request: { type: 'cold' } })
    ),
  })
  feedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual([
    'hot',
    'hot',
    'hot',
    'cold',
    'cold',
    'cold',
  ])
})

test('interleave', () => {
  const actual: string[] = []
  const { addThreads, thread, sync, trigger, feedback, loop } = bProgram()
  addThreads({
    addHot: thread(
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } })
    ),
    addCold: thread(
      sync({ request: { type: 'cold' } }),
      sync({ request: { type: 'cold' } }),
      sync({ request: { type: 'cold' } })
    ),
    mixHotCold: loop([
      sync({
        waitFor: { type: 'hot' },
        block: { type: 'cold' },
      }),
      sync({
        waitFor: { type: 'cold' },
        block: { type: 'hot' },
      }),
    ]),
  })
  feedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual([
    'hot',
    'cold',
    'hot',
    'cold',
    'hot',
    'cold',
  ])
})

test('logging', () => {
  const logs: Parameters<DevCallback>[0][] = []
  const { addThreads, thread, sync, trigger, loop } = bProgram({
    dev: msg => logs.push(msg),
  })
  addThreads({
    addHot: thread(
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } })
    ),
    addCold: thread(
      sync({ request: { type: 'cold' } }),
      sync({ request: { type: 'cold' } }),
      sync({ request: { type: 'cold' } })
    ),
    mixHotCold: loop([
      sync({
        waitFor: { type: 'hot' },
        block: { type: 'cold' },
      }),
      sync({
        waitFor: { type: 'cold' },
        block: { type: 'hot' },
      }),
    ]),
  })
  trigger({ type: 'start' })
  expect(logs).toMatchSnapshot()
})
