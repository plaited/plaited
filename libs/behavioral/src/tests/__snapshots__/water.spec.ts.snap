// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`logging 1`] = `
[
  [
    {
      "priority": 0,
      "request": [
        {
          "detail": undefined,
          "type": "start",
        },
      ],
      "selected": "start",
      "thread": "start",
      "trigger": true,
      "waitFor": [
        {
          "cb": [Function],
          "type": "",
        },
      ],
    },
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "cold",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "hot",
        },
      ],
    },
  ],
  [
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "selected": "hot",
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "cold",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "hot",
        },
      ],
    },
  ],
  [
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "selected": "cold",
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "hot",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "cold",
        },
      ],
    },
  ],
  [
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "selected": "hot",
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "cold",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "hot",
        },
      ],
    },
  ],
  [
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "selected": "cold",
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "hot",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "cold",
        },
      ],
    },
  ],
  [
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "selected": "hot",
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "cold",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "hot",
        },
      ],
    },
  ],
  [
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "selected": "cold",
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "hot",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "cold",
        },
      ],
    },
  ],
]
`;
